doc/arch/001-screen-controllers.md

```diff
@@ -0,0 +1,102 @@
+# ADR: Screen Controllers
+
+**Status**
+Accepted
+
+**Context**
+We don’t want to put business logic inside Android framework classes (like an `Activity` or `Fragment`) because those cannot be unit tested. To enable a fast feedback loop (i.e. tests that run on the JVM and not Android VM), we separate screens and controllers using the [MVI architecture](https://medium.com/@ragunathjawahar/mvi-series-a-pragmatic-reactive-architecture-for-android-87ca176387d1) [pattern](https://medium.com/@ragunathjawahar/mvi-series-a-pragmatic-reactive-architecture-for-android-87ca176387d1).
+
+**Decision**
+Every screen has one controller that consumes user events, performs business logic with the help of data repositories and communicates UI changes back to the screen.
+
+User interactions are abstracted inside data classes of type `UiEvent`. These events flow to the controller in the form of RxJava streams.
+
+```kotlin
+// Create the UsernameTextChanged event by listening to the EditText
+RxTextView
+  .textChanges(usernameEditText)
+  .map { text -> UsernameTextChanged(text) }
+
+// Event
+data class UsernameTextChanged(text: String) : UiEvent
+```
+
+The screen sends a single stream of UI events to the controller and gets back a transformed stream of UI changes. The flow of data is unidirectional. To merge multiple streams into one, RxJava’s `merge()`  operator is used.
+
+```kotlin
+// Login screen
+Observable.merge(usernameChanges(), passwordChanges(), submitClicks())
+  .compose(controller)
+  .takeUntil(screenDestroy)
+  .subscribe { uiChange -> uiChange(this) }
+```
+
+In the controller, UI events are transformed as per the business logic and UI changes are sent back to the screen.
+
+```kotlin
+typealias Ui = LoginScreen
+typealias UiChange = (LoginScreen) -> Unit
+
+class LoginScreenController : ObservableTransfer<UiEvent, UiChange>() {
+
+  fun apply(events: Observable<UiEvent>): Observable<UiChange> {
+    val replayedEvents = events.replay().refCount()
+    return Observable.merge(
+        enableSubmitButton(replayedEvents),
+        loginUser(replayedEvents))
+  }
+
+  fun enableSubmitButton(events: Observable<UiEvent>): Observable<UiChange>() {
+    return events.ofType<UsernameTextChanged>
+      .map { isValidUsername(it.text) }
+      .map { isValid ->
+          { ui: Ui -> ui.setSubmitButtonEnabled(isValid) }
+      }
+  }
+
+  fun loginOnSubmit(events: Observable<UiEvent>): Observable<UiChange>() {
+    ...
+  }
+}
+```
+
+**Consequences**
+Because all events are replayed from the beginning on subscription, `UiChange` functions can only subscribe to events exactly once, and immediately when the controller is constructed.
+
+For instance, subscribing to the event stream in a `flatMap` will result in the events getting replayed every time `flatMap`'s upstream emits:
```

Comment:
I am not sure if I understand this entirely.
Are we pointing to the fact that `take(1)` will end the stream and subsequent subscriptions will not receive any UI events ?
If that's the case, should we not include our solution to use `replay().refCount()` ?
Also, I find the above statement a little confusing.
I can discuss this in-person.

Reply:
> Are we pointing to the fact that `take(1)` will end the stream and subsequent subscriptions will not receive any UI events ?

No. The problem is that because events get replayed from the start when they're subscribed, `take(1)` will emit the *first* item and not the *latest*. 

I can sit with you an explain these two examples.